pkg/forward/ringbuffer.go

```diff
@@ -0,0 +1,87 @@
+package forward
+
+import (
+	"sync"
+)
+
+// RingBuffer is a fixed-length ring buffer. It can be used by a forwarder, to 'drop' messages instead of applying backpressure. See Issue #15
+type RingBuffer struct {
+	max int
+
+	buf   []string
+	first int          // the index of the first item in the buffer
+	last  int          // the index of the last item in the buffer
+	ch    chan string  // the channel is used to block on the Get method whenever data is unavailable
+	mutex sync.RWMutex // synchronizes changes to buf, first, last
+}
+
+// Put() processes the record without blocking.
+// It's behaviour varies depending on the state of the buffer and any blocking Get() invocations
+// It either sends the record over the channel, adds it to the buffer, or drops the record if the buffer is full.
+func (b *RingBuffer) Put(record string) {
+	select {
+	case b.ch <- record: // no need to even add to the buffer if something is blocking on the channel
```

Comment:
I don't like this dual-mode implementation, where the records follow one datapath if there happens to be a waiting reader, and another if there doesn't. The Get method in particular is highly complected as a consequence of this design. Instead, all records should flow through the buffer in all cases.